Add mode option to open/create

dubiousjim opened the pull request:
This is part of a series of PRs towards #4017.

This PR adds a `mode` option to open and create, paralleling the similar options for mkdir, make_temp, and truncate. (It renames the existing `mode` parameter for open and create to `openMode`.)

As with mkdir, makeTemp*, and truncate, this argument is only used when a new file is created. Defaults to 0o666 (the process's umask will then be applied).
    
Currently we throw if `mode` is specified but a read-only opening mode was requested.

dubiousjim commented:
We need to decide how to handle `mode` on Windows: NOOP? or throw? The codebase doesn't seem to be consistent on this.

Also I need to (and will) add a unit test for this new functionality.

ry commented on cli/js/files_test.ts:
```diff
@@ -74,6 +74,36 @@ unitTest(async function readerToAsyncIterator(): Promise<void> {
+    // assertEquals(pathInfo.mode!, 0o626 & ~Deno.umask());
+    assertEquals(pathInfo.mode! & 0o777, 0o604); // assume umask 0o022
+  }
+);
```
These tests seem good to me.

ry commented on cli/js/files.ts:
```diff
@@ -26,71 +26,84 @@ let OP_WRITE = -1;
+  path: string,
+  openMode?: OpenMode,
+  mode?: number
+): Promise<File>;
```
This signature seems not ok to me, also not tested. 
https://deno.land/std/style_guide.md#exported-functions-max-2-args-put-the-rest-into-an-options-object

dubiousjim replied:
Sure, just let me know whether you prefer (1) this signature to go back to omitting `mode?: number`, or (2) I add a test for this signature.

hayd replied:
What about `mode: number | OpenMode` ?

ry replied:
@dubiousjim omit this signature

dubiousjim commented:
This isn't ready to be merged... [two issues](https://github.com/denoland/deno/pull/4286#discussion_r390191212) need to be resolved. On noop vs throw (see #4306), @ry prefers noop but @kitsonk prefers throwing.

kitsonk commented:
@kitsonk only has opinions... @ry is the design authority. 😄

dubiousjim commented:
This could be ready to go: I removed the signature @ry requested, and resolved the ambiguous comment. As with #4287 and #4286, I implemented things by throwing on Windows if an explicit `mode` was provided. For two reasons: 1. it's easier to reverse that and switch to noop than the other way, and 2. the recent umask(mode) PR went with throwing on Windows, and as I said in #4306, I hoped for a consistent policy, to guide how to proceed when implementing new functionality. But it'd be easy, and I'm happy, to change this to make WIndows silently ignore an explicit `mode` if that's what's wanted.

ry commented on cli/js/files.ts:
```diff
@@ -92,8 +116,14 @@ export function createSync(path: string): File {
  */
-export function create(path: string): Promise<File> {
-  return open(path, "w+");
+export function create(path: string, mode?: number): Promise<File> {
```
create should not have a mode argument: https://golang.org/pkg/os/#Create

It should default to 0666 before umask.

ry commented on cli/js/ops/fs/open.ts:
```diff
@@ -29,9 +29,14 @@ export interface OpenOptions {
+  /** Permissions to use if creating the file (defaults to `0o666`, before
+   * the process's umask).
+   * It's an error to specify mode without also setting create or createNew to `true`.
+   * Raises on Windows. */
```
Mode should be ignored on windows - and not result in error.

@piscisaureus correct?

ry commented on cli/js/tests/files_test.ts:
```diff
@@ -74,6 +74,42 @@ unitTest(async function readerToAsyncIterator(): Promise<void> {
 
+unitTest(
+  {
+    skip: Deno.build.os === "win",
```
Don't skip windows, only skip last assert on windows.

ry commented on cli/js/ops/fs/open.ts:
```diff
@@ -29,9 +29,14 @@ export interface OpenOptions {
 }
 
-/** A set of string literals which specify the open mode of a file.
+/** A set of string literals which specify the openMode of a file.
```
s/specify the openMode of a file/specify how to open a file/

ry commented on cli/ops/fs.rs:
```diff
@@ -15,10 +15,10 @@ use std::fs;
 use std::path::Path;
 use std::time::UNIX_EPOCH;
-use tokio;
+use tokio::fs as tokio_fs;
```
tokio::fs is one character shorter than the tokio_fs. This alias just adds complexity (admittedly a very small amount, but still more complex than not adding alias.)

dubiousjim replied:
Sure, will delete. It made more sense in the context of the whole batch of commits I'm now pushing piecemeal, since there are some functions using std::fs, some using tokio::fs, and some using cli/fs.rs. I thought it'd be easier to keep track by referring to them as fs, tokio_fs, and deno_fs, respectively. But it's not clear that things will look the same in the end anyway, as these PRs get refined and merged.

Will push commits making all these revisions shortly.

dubiousjim commented:
Pushed 5 changes:
1. removed deno_typescript stuff from branch
2. tokio_fs -> tokio::fs
3. made open/openSync ignore mode option on Windows, rather than throwing. Adjusted tests to correspond.
4. Fixed doc/comment on OpenMode as @ry proposed
5. removed mode argument from create/createSync

ry commented on cli/ops/fs.rs:
```diff
@@ -84,18 +99,29 @@ fn op_open(
+    if gave_mode && !(options.create || options.create_new) {
+      return Err(OpError::type_error(
+        "specified mode without allowing file creation".to_string(),
+      ));
```
This error handling seems unnecessary. Rust will raise an the error.

dubiousjim replied:
If I remove this check, Rust raises no error for `Deno.openSync(existing_filepath, { create:false, createNew:false, mode:0o600, read:true, write:true});`. See below.

ry replied:
If Rust doesn't raise, why do we need to?

ry commented on cli/ops/fs.rs:
```diff
@@ -84,18 +99,29 @@ fn op_open(
+          return Err(OpError::type_error(
+            "specified mode for read-only open".to_string(),
+          ));
+        }
```
Unnecessary. Let lower level code handle it.

dubiousjim replied:
If I remove this check, Rust raises no error for `Deno.openSync(existing_filepath, { mode:0o600, read:true, write:false});` ... but this would also be caught by the previous error-check, so that one may be enough. (If one attempts to open with `{create: true, write: false}`, we already catch that and report an error in `cli/js/files.ts`.)

I was thinking that supplying a `mode` option to `open` or `openSync` when neither `create` nor `createNew` was true was most likely to be a programming error, and so it'd be useful to throw, rather than just silently ignoring the `mode` option. But if you think the latter is better, it's easy to implement. I'll push that change. If in light of this info you think differently, you can just revert it.

dubiousjim commented:
Hi @ry, I had the impression you thought this PR would be ready to go after the changes most recently pushed. It's holding up other PRs I have queued --- both towards my series #4017 of various fs stuff, and some work I've just done on #4188. (The latter required a bunch of refactoring of cli/fs.rs and cli/ops/fs.rs, and I'd rather not have to work around this one still being in limbo.)

As this branch now stands, all the errors for specifying mode on Windows, or specifying mode when one forbade the creation of a new file, are omitted. I just asked in a comment above whether it might after all be helpful to include the latter errors. But am happy to go along with whatever your considered view is there.
